BibTeX.js

```diff
@@ -889,30 +890,67 @@ function beginRecord(type, closeChar) {
 }
 
 function doImport() {
-	var read = "", text = "", recordCloseElement = false;
+	if (typeof Promise == 'undefined') {
+		readString(
+			function () {},
+			function (e) {
+				throw e;
+			}
+		);
+	}
+	else {
+		return new Promise(function (resolve, reject) {
+			readString(resolve, reject);
+		});
+	}
+}
+
+function readString(resolve, reject) {
+	var read = "";
 	var type = false;
 	
-	while(read = Zotero.read(1)) {
-		if(read == "@") {
-			type = "";
-		} else if(type !== false) {
-			if(type == "comment") {
-				processComment();
-				type = false;
-			} else if(read == "{") {		// possible open character
-				beginRecord(type, "}");
-				type = false;
-			} else if(read == "(") {		// possible open character
-				beginRecord(type, ")");
-				type = false;
-			} else if(/[a-zA-Z0-9-_]/.test(read)) {
-				type += read;
+	var next = function () {
+		readString(resolve, reject);
+	};
+	
+	try {
+		while (read = Zotero.read(1)) {
+			if(read == "@") {
+				type = "";
+			} else if(type !== false) {
+				if(type == "comment") {
+					processComment();
+					type = false;
+				} else if(read == "{") {		// possible open character
+					// This might return a promise if an item was saved
+					// TODO: When 5.0-only, make sure this always returns a promise
+					let maybePromise = beginRecord(type, "}");
```

Comment:
Can/should we use `let` in translator code which has to run in various browsers?